src/java/org/apache/cassandra/cql3/restrictions/SimpleRestriction.java

```diff
@@ -0,0 +1,374 @@
+        return columnsExpression.isColumnLevelExpression();
+    }
+
+    public Operator operator()
+    {
+        return operator;
+    }
+
+    @Override
+    public boolean isANN()
+    {
+        return operator == Operator.ANN;
+    }
+
+    @Override
+    public boolean isEQ()
+    {
+        return operator == Operator.EQ;
+    }
+
+    @Override
+    public boolean isSlice()
+    {
+        return operator.isSlice();
+    }
+
+    @Override
+    public boolean isIN()
+    {
+        return operator == Operator.IN;
+    }
+
+    /**
+     * Checks if this restriction operator is a CONTAINS, CONTAINS_KEY or is an equality on a map element.
+     * @return {@code true} if the restriction operator is one of the contains operations, {@code false} otherwise.
+     */
+    public boolean isContains()
+    {
+        return operator == Operator.CONTAINS
+               || operator == Operator.CONTAINS_KEY
+               || columnsExpression.kind() == ColumnsExpression.Kind.MAP_ELEMENT;
+    }
+
+    @Override
+    public boolean needsFilteringOrIndexing()
+    {
+        // The need for filtering or indexing is a combination of columns expression type and operator
+        // Therefore, we have to take both into account.
+        return columnsExpression.kind() == ColumnsExpression.Kind.MAP_ELEMENT
+               || operator.requiresFilteringOrIndexingFor(columnsExpression.columnsKind());
+    }
+
+    @Override
+    public void addFunctionsTo(List<Function> functions)
+    {
+        columnsExpression.addFunctionsTo(functions);
+        values.addFunctionsTo(functions);
+    }
+
+    @Override
+    public boolean needsFiltering(Index.Group indexGroup)
+    {
+        for (ColumnMetadata column : columns())
+        {
+            if (!isSupportedBy(indexGroup.getIndexes(), column))
+                return true;
+        }
+        return false;
+    }
+
+    private boolean isSupportedBy(Iterable<Index> indexes, ColumnMetadata column)
+    {
+        if (isOnToken())
+            return false;
+
+        for (Index index : indexes)
+        {
+            if (index.supportsExpression(column, operator))
+                return true;
+        }
+        return false;
+    }
+
+    @Override
+    public Index findSupportingIndex(Iterable<Index> indexes)
+    {
+        if (isOnToken())
+            return null;
+
+        for (Index index : indexes)
+            if (isSupportedBy(index))
+                return index;
+        return null;
+    }
+
+    @Override
+    public boolean isSupportedBy(Index index)
+    {
+        if (isOnToken())
+            return false;
+
+        for (ColumnMetadata column : columns())
+        {
+            if (index.supportsExpression(column, operator))
+                return true;
+        }
+        return false;
+    }
+
+    @Override
+    public List<ClusteringElements> values(QueryOptions options)
+    {
+        assert operator == Operator.EQ || operator == Operator.IN || operator == Operator.ANN;
```

Comment:
```suggestion
        assert operator == Operator.EQ || operator == Operator.IN;
```
It seems the `bindAndGetClustering` is replicating `appendTo` which was unsupported operation for `ANN` according to ex `SingleColumnRestriction`?

Reply:
On a second look - it seems `appendTo` and `value` in ANN Restriction are now `values` method.
We will use the new method, while before, we were throwing `UnsupportedOperation` when we see ANN. 
https://github.com/apache/cassandra/pull/3095/commits/b17c68b69bc712118b330a4188e775d8e53f5003#diff-ff9b594718890837b5d63cfe718c916233be6c9b891de192ad5e7ece33ccfa47L110-R107

Reply:
We use the `values` in `StorageAttachedIndex` but we are not calling it from the other path.

Reply:
`We use the values in StorageAttachedIndex`
The old` value` method - yes, the new `values` one - no. We call it at other places where we had `appendTo` being called before, but if you say at those places we would never hit the ANN override, I guess that is ok.